engine/engine_manager.go

```diff
@@ -318,6 +350,48 @@ func (em *Manager) EngineLog(req *rpc.LogRequest, srv rpc.EngineManagerService_E
 	return nil
 }
 
+func (em *Manager) EngineStream(req *empty.Empty, srv rpc.EngineManagerService_EngineStreamServer) error {
+	responseChan := make(chan *rpc.EngineResponse)
+	em.lock.Lock()
+	em.rpcStreams[responseChan] = struct{}{}
+	em.lock.Unlock()
+	defer func() {
+		em.lock.Lock()
+		delete(em.rpcStreams, responseChan)
+		em.lock.Unlock()
+	}()
+
+	var initialEngineState []*rpc.EngineResponse
+	em.lock.RLock()
+	for _, e := range em.engineLaunchers {
+		e.lock.RLock()
+		name := e.currentEngine.EngineName
+		e.lock.RUnlock()
+		response, err := em.processManager.ProcessGet(nil, &rpc.ProcessGetRequest{
+			Name: name,
+		})
+		if err != nil {
+			return err
+		}
+		initialEngineState = append(initialEngineState, e.RPCResponse(response))
+	}
+	em.lock.RUnlock()
+
+	for _, resp := range initialEngineState {
+		if err := srv.Send(resp); err != nil {
```

Comment:
It's maybe faster if we stream a list instead of objects one by one. Depends on the protocol, there can be less overhead in transmitting.

Reply:
Would it be faster to stream a `list` of `objects`? I feel like in the case of there being numerous different `objects` updating within a short time-frame, sending multiple `list` responses to the `Client` and having to parse through multiple `lists` of all `objects` versus just the updated `objects` would be slower on the `client` end.

We could collect only the updated `objects` and send that, but then we'd need to specify some time interval to wait for updated `objects` and then send versus just sending off any updates immediately.

Using the `list` only to report the initial `State` from the `Server` doesn't make sense.

Between the current approach and collecting updates within a certain time interval to send off, which do you think makes more sense?

Reply:
What I was referring to was the extra protocol wrapper needed for each object. I don't have a definitive answer but normally one big package took less time transmitting in the network, though it would depend on the protocol implementation. 

I am OK with the current approach. If we benchmark it later and found some room to improve, we can always do it later.

Reply:
I'll leave the current implementation of this as in then, and then revisit it if the benchmarks suggest room for improvement as you say.